File: kotlin/src/main/kotlin/com/solana/mpp/client/PaymentChannels.kt
package com.solana.mpp.client

import com.solana.mpp.crypto.AccountMeta
import com.solana.mpp.crypto.Blake3
import com.solana.mpp.crypto.Instruction
import com.solana.mpp.crypto.Pda
import com.solana.mpp.crypto.Programs
import com.solana.mpp.crypto.PublicKey

/**
 * Typed helpers for the on-chain payment-channels program used by the MPP
 * session intent: channel PDA derivation, associated-token derivation, the
 * blake3 distribution hash, the Borsh voucher signing bytes, the Ed25519
 * precompile verify instruction, and the Open / TopUp instruction builders.
 *
 * Byte layouts mirror `rust/crates/mpp/src/program/payment_channels.rs`
 * exactly so cross-language vouchers and PDAs match. The voucher signing
 * bytes are channelId(32) || cumulativeAmount(u64 LE, 8) ||
 * expiresAt(i64 LE, 8) = 48 bytes, signed with Ed25519.
 */
object PaymentChannels {
    /** Canonical payment-channels program ID deployed to Surfnet. */
    const val PAYMENT_CHANNELS_PROGRAM_ID = "GuoKrzaBiZnW5DvJ3yZVE7xHqbcBvaX9SH6P6Cn9gNvc"

    /** Channel PDA seed prefix. */
    const val CHANNEL_SEED = "channel"

    /** Event authority PDA seed prefix. */
    const val EVENT_AUTHORITY_SEED = "event_authority"

    /** Ed25519 signature-verify precompile program ID. */
    const val ED25519_PROGRAM_ID = "Ed25519SigVerify111111111111111111111111111"

    /** Instructions sysvar account ID. */
    const val INSTRUCTIONS_SYSVAR_ID = "Sysvar1nstructions1111111111111111111111111"

    /** Rent sysvar account ID. */
    const val RENT_SYSVAR_ID = "SysvarRent111111111111111111111111111111111"

    /** Length of the Borsh voucher signing message. */
    const val VOUCHER_MESSAGE_LEN = 48

    // Anchor instruction discriminators for the payment-channels program.
    private const val OPEN_DISCRIMINATOR: Byte = 1
    private const val TOP_UP_DISCRIMINATOR: Byte = 3

    /** A single split recipient with a basis-point share. */
    data class Distribution(val recipient: PublicKey, val bps: Int) {
        init {
            require(bps in 0..0xffff) { "bps must fit in u16 (got $bps)" }
        }
    }

    /** Everything needed to derive the channel addresses and build Open. */
    data class OpenChannelParams(
        val payer: PublicKey,
        val payee: PublicKey,
        val mint: PublicKey,
        val authorizedSigner: PublicKey,
        val salt: Long,
        val deposit: Long,
        val gracePeriod: Int,
        val recipients: List<Distribution>,
        val tokenProgram: PublicKey,
        val programId: PublicKey,
    )

    /** Derived accounts for a channel open. */
    data class ChannelAddresses(
        val channel: PublicKey,
        val payerTokenAccount: PublicKey,
        val channelTokenAccount: PublicKey,
        val eventAuthority: PublicKey,
    )

    /** Returns the canonical payment-channels program ID. */
    fun defaultProgramId(): PublicKey = PublicKey.fromBase58(PAYMENT_CHANNELS_PROGRAM_ID)

    /**
     * Derives the channel PDA. Seed order mirrors the Rust spine:
     * "channel" || payer || payee || mint || authorizedSigner || salt(u64 LE).
     */
    fun findChannelPda(
        payer: PublicKey,
        payee: PublicKey,
        mint: PublicKey,
        authorizedSigner: PublicKey,
        salt: Long,
        programId: PublicKey,
    ): Pair<PublicKey, Int> =
        Pda.findProgramAddress(
            listOf(
                CHANNEL_SEED.encodeToByteArray(),
                payer.bytes,
                payee.bytes,
                mint.bytes,
                authorizedSigner.bytes,
                u64LE(salt),
            ),
            programId,
        )

    /** Derives the event authority PDA for the program. */
    fun findEventAuthorityPda(programId: PublicKey): Pair<PublicKey, Int> =
        Pda.findProgramAddress(listOf(EVENT_AUTHORITY_SEED.encodeToByteArray()), programId)

    /**
     * Derives the ATA for the given owner, mint, and token program. Seed
     * order is owner || tokenProgram || mint under the associated-token
     * program, matching the Rust spine.
     */
    fun findAssociatedTokenAddress(owner: PublicKey, mint: PublicKey, tokenProgram: PublicKey): PublicKey =
        Pda.associatedTokenAddress(owner, mint, tokenProgram)

    /** Derives the channel PDA plus the payer/channel ATAs and event authority. */
    fun deriveChannelAddresses(params: OpenChannelParams): ChannelAddresses {
        val (channel, _) = findChannelPda(
            params.payer, params.payee, params.mint, params.authorizedSigner, params.salt, params.programId,
        )
        val payerAta = findAssociatedTokenAddress(params.payer, params.mint, params.tokenProgram)
        val channelAta = findAssociatedTokenAddress(channel, params.mint, params.tokenProgram)
        val (eventAuthority, _) = findEventAuthorityPda(params.programId)
        return ChannelAddresses(
            channel = channel,
            payerTokenAccount = payerAta,
            channelTokenAccount = channelAta,
            eventAuthority = eventAuthority,
        )
    }

    /**
     * Returns the 32 byte blake3 distribution hash committed at channel open.
     * Preimage is count(u32 LE) followed by each recipient(32) || bps(u16 LE),
     * matching `distribution_hash` in the Rust spine.
     */
    fun distributionHash(recipients: List<Distribution>): ByteArray {
        val hasher = Blake3()
        hasher.update(u32LE(recipients.size))
        for (r in recipients) {
            hasher.update(r.recipient.bytes)
            hasher.update(u16LE(r.bps))
        }
        return hasher.finalize()
    }

    /**
     * Returns the Borsh VoucherArgs bytes signed by Ed25519:
     * channelId(32) || cumulativeAmount(u64 LE) || expiresAt(i64 LE).
     */
    fun voucherMessageBytes(channelId: PublicKey, cumulativeAmount: Long, expiresAt: Long): ByteArray {
        val out = ByteArray(VOUCHER_MESSAGE_LEN)
        System.arraycopy(channelId.bytes, 0, out, 0, 32)
        writeU64LE(out, 32, cumulativeAmount)
        writeU64LE(out, 40, expiresAt)
        return out
    }

    /**
     * Builds the Ed25519 precompile verify instruction over the voucher
     * message. Layout mirrors `build_ed25519_verify_instruction` in the
     * Rust spine.
     */
    fun buildEd25519VerifyInstruction(
        authorizedSigner: PublicKey,
        signature: ByteArray,
        message: ByteArray,
    ): Instruction {
        require(signature.size == 64) { "ed25519 signature must be 64 bytes (got ${signature.size})" }
        require(message.size <= 0xffff) { "voucher message too large for ed25519 instruction" }
        val publicKeyOffset = 16
        val signatureOffset = publicKeyOffset + 32
        val messageDataOffset = signatureOffset + 64
        val currentInstruction = 0xffff
        val messageDataSize = message.size

        val data = ArrayList<Byte>(messageDataOffset + message.size)
        data.add(1) // count
        data.add(0) // padding
        appendU16LE(data, signatureOffset)
        appendU16LE(data, currentInstruction)
        appendU16LE(data, publicKeyOffset)
        appendU16LE(data, currentInstruction)
        appendU16LE(data, messageDataOffset)
        appendU16LE(data, messageDataSize)
        appendU16LE(data, currentInstruction)
        for (b in authorizedSigner.bytes) data.add(b)
        for (b in signature) data.add(b)
        for (b in message) data.add(b)

        return Instruction(
            programId = ED25519_PROGRAM_ID,
            accounts = emptyList(),
            data = data.toByteArray(),
        )
    }

    /**
     * Builds the payment-channels Open instruction. The account order and
     * Borsh arg encoding mirror the generated Codama client used by the Rust
     * spine (discriminator 1, 13 accounts).
     */
    fun buildOpenInstruction(params: OpenChannelParams): Instruction {
        val addresses = deriveChannelAddresses(params)
        val data = ByteArray(1) { OPEN_DISCRIMINATOR } + encodeOpenArgs(params)
        val accounts = listOf(
            AccountMeta.writable(params.payer.toBase58(), signer = true),
            AccountMeta.readOnly(params.payee.toBase58()),
            AccountMeta.readOnly(params.mint.toBase58()),
            AccountMeta.readOnly(params.authorizedSigner.toBase58()),
            AccountMeta.writable(addresses.channel.toBase58()),
            AccountMeta.writable(addresses.payerTokenAccount.toBase58()),
            AccountMeta.writable(addresses.channelTokenAccount.toBase58()),
            AccountMeta.readOnly(params.tokenProgram.toBase58()),
            AccountMeta.readOnly(Programs.SYSTEM_PROGRAM),
            AccountMeta.readOnly(RENT_SYSVAR_ID),
            AccountMeta.readOnly(Programs.ASSOCIATED_TOKEN_PROGRAM),
            AccountMeta.readOnly(addresses.eventAuthority.toBase58()),
            AccountMeta.readOnly(params.programId.toBase58()),
        )
        return Instruction(programId = params.programId.toBase58(), accounts = accounts, data = data)
    }

    /**
     * Builds the payment-channels TopUp instruction (discriminator 3, 6
     * accounts) raising a channel's deposit by amount.
     */
    fun buildTopUpInstruction(
        payer: PublicKey,
        channel: PublicKey,
        mint: PublicKey,
        amount: Long,
        tokenProgram: PublicKey,
        programId: PublicKey,
    ): Instruction {
        val payerAta = findAssociatedTokenAddress(payer, mint, tokenProgram)
        val channelAta = findAssociatedTokenAddress(channel, mint, tokenProgram)
        val data = ByteArray(1) { TOP_UP_DISCRIMINATOR } + u64LE(amount)
        val accounts = listOf(
            AccountMeta.writable(payer.toBase58(), signer = true),
            AccountMeta.writable(channel.toBase58()),
            AccountMeta.writable(payerAta.toBase58()),
            AccountMeta.writable(channelAta.toBase58()),
            AccountMeta.readOnly(mint.toBase58()),
            AccountMeta.readOnly(tokenProgram.toBase58()),
        )
        return Instruction(programId = programId.toBase58(), accounts = accounts, data = data)
    }

    /**
     * Borsh-encodes OpenArgs: salt(u64 LE) || deposit(u64 LE) ||
     * gracePeriod(u32 LE) || recipients(vec: u32 LE len, then
     * recipient(32) || bps(u16 LE)).
     */
    private fun encodeOpenArgs(params: OpenChannelParams): ByteArray {
        val out = ArrayList<Byte>(8 + 8 + 4 + 4 + params.recipients.size * 34)
        for (b in u64LE(params.salt)) out.add(b)
        for (b in u64LE(params.deposit)) out.add(b)
        for (b in u32LE(params.gracePeriod)) out.add(b)
        for (b in u32LE(params.recipients.size)) out.add(b)
        for (r in params.recipients) {
            for (b in r.recipient.bytes) out.add(b)
            for (b in u16LE(r.bps)) out.add(b)
        }
        return out.toByteArray()
    }

    private fun u16LE(value: Int): ByteArray =
        byteArrayOf((value and 0xff).toByte(), ((value ushr 8) and 0xff).toByte())

    private fun u32LE(value: Int): ByteArray {
        val out = ByteArray(4)
        for (i in 0 until 4) out[i] = ((value ushr (8 * i)) and 0xff).toByte()
        return out
    }

    private fun u64LE(value: Long): ByteArray {
        val out = ByteArray(8)
        writeU64LE(out, 0, value)
        return out
    }

    private fun writeU64LE(out: ByteArray, offset: Int, value: Long) {
        for (i in 0 until 8) out[offset + i] = ((value ushr (8 * i)) and 0xff).toByte()
    }

    private fun appendU16LE(buf: MutableList<Byte>, value: Int) {
        buf.add((value and 0xff).toByte())
        buf.add(((value ushr 8) and 0xff).toByte())
    }
}

File: go/protocols/mpp/program/payment_channels.go
// Package program carries typed helpers for the on-chain payment-channels
// program used by the MPP session intent: channel PDA derivation,
// associated-token derivation, the blake3 distribution hash, the Borsh
// voucher signing bytes, the Ed25519 precompile verify instruction, and
// the Open / TopUp instruction builders.
//
// Byte layouts mirror rust/crates/mpp/src/program/payment_channels.rs
// exactly so cross-language vouchers and PDAs match. The voucher signing
// bytes are channel_id(32) || cumulative_amount(u64 LE,8) ||
// expires_at(i64 LE,8) = 48 bytes, signed with Ed25519.
package program

import (
	"encoding/binary"
	"fmt"

	solana "github.com/gagliardetto/solana-go"
	"lukechampine.com/blake3"
)

// PaymentChannelsProgramID is the canonical payment-channels program ID
// deployed to Surfnet. Mirrors PAYMENT_CHANNELS_PROGRAM_ID in the Rust spine.
const PaymentChannelsProgramID = "GuoKrzaBiZnW5DvJ3yZVE7xHqbcBvaX9SH6P6Cn9gNvc"

// Seed prefixes and sysvar / precompile IDs used by the program.
const (
	// ChannelSeed is the channel PDA seed prefix.
	ChannelSeed = "channel"
	// EventAuthoritySeed is the event authority PDA seed prefix.
	EventAuthoritySeed = "event_authority"
	// Ed25519ProgramID is the Ed25519 signature-verify precompile program ID.
	Ed25519ProgramID = "Ed25519SigVerify111111111111111111111111111"
	// InstructionsSysvarID is the instructions sysvar account ID.
	InstructionsSysvarID = "Sysvar1nstructions1111111111111111111111111"
	// RentSysvarID is the rent sysvar account ID.
	RentSysvarID = "SysvarRent111111111111111111111111111111111"
)

// VoucherMessageLen is the length of the Borsh voucher signing message.
const VoucherMessageLen = 48

// Distribution is a single split recipient with a basis-point share.
type Distribution struct {
	Recipient solana.PublicKey
	Bps       uint16
}

// DefaultProgramID returns the canonical payment-channels program ID.
func DefaultProgramID() solana.PublicKey {
	return solana.MustPublicKeyFromBase58(PaymentChannelsProgramID)
}

// FindChannelPDA derives the channel PDA. Seed order mirrors the Rust spine:
// "channel" || payer || payee || mint || authorizedSigner || salt(u64 LE).
func FindChannelPDA(
	payer, payee, mint, authorizedSigner solana.PublicKey,
	salt uint64,
	programID solana.PublicKey,
) (solana.PublicKey, uint8, error) {
	saltLE := make([]byte, 8)
	binary.LittleEndian.PutUint64(saltLE, salt)
	pda, bump, err := solana.FindProgramAddress([][]byte{
		[]byte(ChannelSeed),
		payer[:],
		payee[:],
		mint[:],
		authorizedSigner[:],
		saltLE,
	}, programID)
	return pda, bump, err
}

// FindEventAuthorityPDA derives the event authority PDA for the program.
func FindEventAuthorityPDA(programID solana.PublicKey) (solana.PublicKey, uint8, error) {
	return solana.FindProgramAddress([][]byte{[]byte(EventAuthoritySeed)}, programID)
}

// FindAssociatedTokenAddress derives the ATA for the given owner, mint, and
// token program. Seed order is owner || tokenProgram || mint under the
// associated-token program, matching the Rust spine.
func FindAssociatedTokenAddress(owner, mint, tokenProgram solana.PublicKey) (solana.PublicKey, uint8, error) {
	return solana.FindProgramAddress([][]byte{
		owner[:],
		tokenProgram[:],
		mint[:],
	}, solana.SPLAssociatedTokenAccountProgramID)
}

// DistributionHash returns the 32-byte blake3 distribution hash committed
// at channel open. Preimage is count(u32 LE) followed by each
// recipient(32) || bps(u16 LE), matching distribution_hash in the Rust spine.
func DistributionHash(recipients []Distribution) [32]byte {
	hasher := blake3.New(32, nil)
	countLE := make([]byte, 4)
	binary.LittleEndian.PutUint32(countLE, uint32(len(recipients)))
	_, _ = hasher.Write(countLE)
	bpsLE := make([]byte, 2)
	for _, r := range recipients {
		_, _ = hasher.Write(r.Recipient[:])
		binary.LittleEndian.PutUint16(bpsLE, r.Bps)
		_, _ = hasher.Write(bpsLE)
	}
	var out [32]byte
	copy(out[:], hasher.Sum(nil))
	return out
}

// VoucherMessageBytes returns the Borsh VoucherArgs bytes signed by Ed25519:
// channelID(32) || cumulativeAmount(u64 LE) || expiresAt(i64 LE).
func VoucherMessageBytes(channelID solana.PublicKey, cumulativeAmount uint64, expiresAt int64) []byte {
	out := make([]byte, VoucherMessageLen)
	copy(out[:32], channelID[:])
	binary.LittleEndian.PutUint64(out[32:40], cumulativeAmount)
	binary.LittleEndian.PutUint64(out[40:48], uint64(expiresAt))
	return out
}

// BuildEd25519VerifyInstruction builds the Ed25519 precompile verify
// instruction over the voucher message. Layout mirrors
// build_ed25519_verify_instruction in the Rust spine.
func BuildEd25519VerifyInstruction(
	authorizedSigner solana.PublicKey,
	signature [64]byte,
	message []byte,
) (solana.Instruction, error) {
	const publicKeyOffset uint16 = 16
	const signatureOffset uint16 = publicKeyOffset + 32
	const messageDataOffset uint16 = signatureOffset + 64
	const currentInstruction uint16 = 0xFFFF
	if len(message) > 0xFFFF {
		return nil, fmt.Errorf("voucher message too large for ed25519 instruction")
	}
	messageDataSize := uint16(len(message))

	data := make([]byte, 0, int(messageDataOffset)+len(message))
	data = append(data, 1, 0) // count=1, padding=0
	data = appendU16LE(data, signatureOffset)
	data = appendU16LE(data, currentInstruction)
	data = appendU16LE(data, publicKeyOffset)
	data = appendU16LE(data, currentInstruction)
	data = appendU16LE(data, messageDataOffset)
	data = appendU16LE(data, messageDataSize)
	data = appendU16LE(data, currentInstruction)
	data = append(data, authorizedSigner[:]...)
	data = append(data, signature[:]...)
	data = append(data, message...)

	return solana.NewInstruction(
		solana.MustPublicKeyFromBase58(Ed25519ProgramID),
		solana.AccountMetaSlice{},
		data,
	), nil
}

func appendU16LE(buf []byte, value uint16) []byte {
	return append(buf, byte(value), byte(value>>8))
}

// Anchor instruction discriminators for the payment-channels program.
const (
	openDiscriminator  byte = 1
	topUpDiscriminator byte = 3
)

// OpenChannelParams carries everything needed to derive the channel
// addresses and build the Open instruction.
type OpenChannelParams struct {
	Payer            solana.PublicKey
	Payee            solana.PublicKey
	Mint             solana.PublicKey
	AuthorizedSigner solana.PublicKey
	Salt             uint64
	Deposit          uint64
	GracePeriod      uint32
	Recipients       []Distribution
	TokenProgram     solana.PublicKey
	ProgramID        solana.PublicKey
}

// ChannelAddresses holds the derived accounts for a channel open.
type ChannelAddresses struct {
	Channel             solana.PublicKey
	PayerTokenAccount   solana.PublicKey
	ChannelTokenAccount solana.PublicKey
	EventAuthority      solana.PublicKey
}

// DeriveChannelAddresses derives the channel PDA, the payer and channel
// associated token accounts, and the event authority PDA.
func DeriveChannelAddresses(params OpenChannelParams) (ChannelAddresses, error) {
	channel, _, err := FindChannelPDA(params.Payer, params.Payee, params.Mint, params.AuthorizedSigner, params.Salt, params.ProgramID)
	if err != nil {
		return ChannelAddresses{}, err
	}
	payerATA, _, err := FindAssociatedTokenAddress(params.Payer, params.Mint, params.TokenProgram)
	if err != nil {
		return ChannelAddresses{}, err
	}
	channelATA, _, err := FindAssociatedTokenAddress(channel, params.Mint, params.TokenProgram)
	if err != nil {
		return ChannelAddresses{}, err
	}
	eventAuthority, _, err := FindEventAuthorityPDA(params.ProgramID)
	if err != nil {
		return ChannelAddresses{}, err
	}
	return ChannelAddresses{
		Channel:             channel,
		PayerTokenAccount:   payerATA,
		ChannelTokenAccount: channelATA,
		EventAuthority:      eventAuthority,
	}, nil
}

// encodeOpenArgs Borsh-encodes OpenArgs:
// salt(u64 LE) || deposit(u64 LE) || gracePeriod(u32 LE) ||
// recipients(vec: u32 LE len, then recipient(32) || bps(u16 LE)).
func encodeOpenArgs(params OpenChannelParams) []byte {
	buf := make([]byte, 0, 8+8+4+4+len(params.Recipients)*34)
	u64 := make([]byte, 8)
	binary.LittleEndian.PutUint64(u64, params.Salt)
	buf = append(buf, u64...)
	binary.LittleEndian.PutUint64(u64, params.Deposit)
	buf = append(buf, u64...)
	u32 := make([]byte, 4)
	binary.LittleEndian.PutUint32(u32, params.GracePeriod)
	buf = append(buf, u32...)
	binary.LittleEndian.PutUint32(u32, uint32(len(params.Recipients)))
	buf = append(buf, u32...)
	for _, r := range params.Recipients {
		buf = append(buf, r.Recipient[:]...)
		buf = appendU16LE(buf, r.Bps)
	}
	return buf
}

// BuildOpenInstruction builds the payment-channels Open instruction. The
// account order and Borsh arg encoding mirror the generated Codama client
// used by the Rust spine (discriminator 1, 13 accounts).
func BuildOpenInstruction(params OpenChannelParams) (solana.Instruction, error) {
	addresses, err := DeriveChannelAddresses(params)
	if err != nil {
		return nil, err
	}
	data := append([]byte{openDiscriminator}, encodeOpenArgs(params)...)
	accounts := solana.AccountMetaSlice{
		solana.Meta(params.Payer).WRITE().SIGNER(),
		solana.Meta(params.Payee),
		solana.Meta(params.Mint),
		solana.Meta(params.AuthorizedSigner),
		solana.Meta(addresses.Channel).WRITE(),
		solana.Meta(addresses.PayerTokenAccount).WRITE(),
		solana.Meta(addresses.ChannelTokenAccount).WRITE(),
		solana.Meta(params.TokenProgram),
		solana.Meta(solana.SystemProgramID),
		solana.Meta(solana.MustPublicKeyFromBase58(RentSysvarID)),
		solana.Meta(solana.SPLAssociatedTokenAccountProgramID),
		solana.Meta(addresses.EventAuthority),
		solana.Meta(params.ProgramID),
	}
	return solana.NewInstruction(params.ProgramID, accounts, data), nil
}

// BuildTopUpInstruction builds the payment-channels TopUp instruction
// (discriminator 3, 6 accounts) raising a channel's deposit by amount.
func BuildTopUpInstruction(
	payer, channel, mint solana.PublicKey,
	amount uint64,
	tokenProgram, programID solana.PublicKey,
) (solana.Instruction, error) {
	payerATA, _, err := FindAssociatedTokenAddress(payer, mint, tokenProgram)
	if err != nil {
		return nil, err
	}
	channelATA, _, err := FindAssociatedTokenAddress(channel, mint, tokenProgram)
	if err != nil {
		return nil, err
	}
	u64 := make([]byte, 8)
	binary.LittleEndian.PutUint64(u64, amount)
	data := append([]byte{topUpDiscriminator}, u64...)
	accounts := solana.AccountMetaSlice{
		solana.Meta(payer).WRITE().SIGNER(),
		solana.Meta(channel).WRITE(),
		solana.Meta(payerATA).WRITE(),
		solana.Meta(channelATA).WRITE(),
		solana.Meta(mint),
		solana.Meta(tokenProgram),
	}
	return solana.NewInstruction(programID, accounts, data), nil
}
